elytron/WFCORE-4172-tls-1.3.adoc

```diff
@@ -0,0 +1,108 @@
+= WFCORE-4172 Add support for TLS 1.3
+:author:            Farah Juma
+:email:             fjuma@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+JDK 11 has added support for TLS 1.3. This task is to update Elytron to be able to make use of TLS 1.3.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-4172[WFCORE-4172]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1022[EAP7-1022]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:okotek@redhat.com[Ondrej Kotek]
+
+=== Affected Projects or Components
+
+* WildFly, Security
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* It should be possible to configure a ```server-ssl-context``` and ```client-ssl-context``` in the Elytron subsystem
+using the TLS 1.3 protocol and TLS 1.3 https://tools.ietf.org/html/rfc8446#appendix-B.4[cipher suites]. (Note
+that TLS 1.2 cipher suites cannot be used with TLS 1.3 and vice and versa.)
+
+** A new attribute ```cipher-suite-names``` will be added to both ```server-ssl-context``` and ```client-ssl-context```
```

Comment:
Why is necessary to add new attribute? There is already ```cipher-suite-filter``` attribute on ```*-ssl-context``` which has similar meaning?

Reply:
This is similar to the approach OpenSSL has taken. Because ciphersuites work differently in TLS 1.3, they introduced a new option to specify them and they no longer allow the special characters that were allowed for specifying TLS 1.2 ciphersuites (""+", "!", "-", etc.) for TLS 1.3. See https://wiki.openssl.org/index.php/TLS1.3#Ciphersuites for more details on the separate options.

Reply:
I must admit I don't fully get why they introduced distinct attribute for similar thing. Question is if we should really repeat they decision. 

From user perspective it become confusing. Is there any technical limitation (API requirement) I am missing which prevents of using one attribute? Is it because of scenario when I specify TLSv1.2 in conjuction with TLSv1.3, but cipher suites are disjunct?

What about the future when there will be another TLSv1.3 cipher suites and we will begin to need selector just as on old attribute cipher-suite-filter?

Maybe their parsing logic does not allow them to reuse one attribute, but it seems to me in Elytron case this filter is just to prepare string which will be passed to jdk. I think selector in Elytron is written generic enough to work with TLSv1.3 cipher suites just fine.

@dmlloyd what do you think?

Reply:
The format and parsing for our ```cipher-suite-filter``` attribute is based on the OpenSSL-style pre TLS 1.3 format (see ```CipherSuiteSelector#fromString```). Having two separate attributes allows appropriate default values to be set (e.g., If ```cipher-suite-filter="ECDHE"``` and ```cipher-suite-names``` is left unspecified, any pre-TLS 1.3 cipher suites that support ECDHE will be considered as well as the default cipher suites for TLS 1.3. With only one attribute, we wouldn't know to additionally include the default cipher suites for TLS 1.3.).

Another thing to note is that the format of the cipher suite names has changed significantly. Instead of a cipher suite name being made up of a protocol, key exchange type, authentication type, AEAD cipher mode, and PRF hash algorithm, a TLS 1.3 cipher suite name is now made up of just the protocol, AEAD cipher mode, and HKDF hash algorithm. So some keywords that were previously used to match pre-TLS 1.3 cipher suite names no longer make sense for TLS 1.3.

Also since there are only 5 cipher suites for TLS 1.3 and since TLS 1.2 cipher suites cannot be used with TLS 1.3 and vice versa, I think it makes sense to have the additional attribute. A user likely would not even need to set this attribute explicitly since it already defaults to 3 out of the 5 TLS 1.3 cipher suites.

Reply:
From user perspective, having two attributes is confusing. What about an alternative approach with just one attribute? The TLS 1.3 cipher suites could be part of the original mechanism database (the format is the same). Any special treatment with a TLS 1.3 cipher suite can be enabled by the suite affiliation with the protocol (which is set in the database) and by using new TLS 1.3 related keywords. For example `cipher-suite-filter="TLS1.3_DEFAULT:ECDHE"` would pick the 3 default TLS 1.3 cipher suites and any pre-TLS 1.3 cipher suites that support ECDHE. Are there any obstacles/disadvantages?

@fjuma @darranl WDYT, please?

Reply:
@OndrejKotek One thing to note is that the keywords we currently use in our ```cipher-suite-filter``` are OpenSSL keywords. So introducing new keywords like ```TLS1.3_DEFAULT``` may be confusing. Another thing to note is that for ```cipher-suite-filter``` configuration that doesn't match the TLS 1.3 cipher suites, TLS 1.3 would not get automatically used by default if available without explicitly specifying a special keyword as you mentioned.

Reply:
I think mixing TLS 1.2 and 1.3 selectors might get very confusing given how different they are.  I agree with Farah and the OpenSSL authors that it is best to keep them separate for now.

Reply:
I think the second attribute is going to make sense and be mutually exclusive with the original attribute, by the sounds of it once we are ready to push all users to use TLS 1.3 it sounds like the original attribute would end up being deprecated.

The problem we have here is the two attributes have completely different formats, this really means tooling will need to treat each of them differently and each will be documented differently - validation for each would also be different.

If these two attributes were combined into one we don't have any way to detect the format in use to validate and make use of the value, we can't even rely on the protocol being specified on the SSLContext as alternative implementations may use different names.

I can see where the comments are coming from related to having two related attributes but in this case it does seem it will be a necessity.

Reply:
Thank you for your clarifications and confirmations. I have no more suggestions or objections.